src/main/java/com/enonic/xp/app/welcome/WelcomePageScriptBean.java

```diff
@@ -230,18 +325,110 @@ public String getStatisticsApiUrl()
         return createUrl( jettyConfigServiceSupplier.get().getHttpMonitorPort() );
     }
 
+    private ApplicationInstallResultJson installApplication( final URL url, final byte[] sha512 )
+    {
+        final ApplicationInstallResultJson result = new ApplicationInstallResultJson();
+
+        try
+        {
+            ContextBuilder.from( ContextAccessor.current() ).authInfo(
+                AuthenticationInfo.create().principals( RoleKeys.ADMIN ).user( SUPER_USER ).build() ).build().runWith( () -> {
+
+                final Application application = this.applicationServiceSupplier.get().installGlobalApplication( url, sha512 );
+
+                result.setApplication( toApplicationJson( application ) );
+
+            } );
+
+        }
+        catch ( Exception e )
+        {
+            final String failure = "Failed to process application from " + url;
+            LOG.error( failure, e );
+
+            result.setFailure( failure );
+        }
+        return result;
+    }
+
+    private ApplicationJson toApplicationJson( final Application application )
+    {
+        ApplicationKey applicationKey = application.getKey();
+        ApplicationJson.Builder builder =
+            ApplicationJson.create().application( application ).description( getApplicationDescription( applicationKey ) ).iconAsBase64(
+                getApplicationIconAsBase64( applicationKey ) );
+
+        Resource resource = resourceServiceSupplier.get().getResource( ResourceKey.from( applicationKey, "/webapp/webapp.js" ) );
+        if ( resource.exists() )
+        {
+            String webappUrl = ServletRequestUrlHelper.createUri( ServletRequestHolder.getRequest(), "/webapp/" + application.getKey() );
+            builder.webappUrl( webappUrl );
+        }
+
+        List<String> adminToolsUris = getApplicationAdminToolsUris( applicationKey );
+        adminToolsUris.forEach( builder::addAdminToolsUrl );
+
+        return builder.build();
+    }
+
+    private <V> V lock( Object key, Callable<V> callable )
```

Comment:
too much buzz, no gain. remove this lock as it does not protect anything. Even if it is a copy from somewhere else, don't repeat the same mistake.